### ISSUE TITLE
SWIG does not respect move constructor for the returning value

### ISSUE THREAD

**Issue by luav** [swig/swig#1044]
Return by value with defined move constructor and deleted copy constructor fails on compilation because the constructor semantic is not respected.  
SWIG interface file:
```c++
%module vector_mbmvctor
%{
#include <utility>  // move()
%}

%inline {
struct Member {
	int  val;
	
	Member(): val(0)  {}
	
	Member(const Member&)=delete;
	Member(Member&&)=default;
	
	Member& operator=(const Member&)=delete;
	Member& operator=(Member&&)=default;
	
	Member release()  { return std::move(*this); }
};
}
```
The generated wrapper for `Member release()`  is:
```c++
SWIGINTERN PyObject *_wrap_Member_release(PyObject *SWIGUNUSEDPARM(self), PyObject *args) {
  ...
  result = (arg1)->release();
  resultobj = SWIG_NewPointerObj((new Member(static_cast< const Member& >(result))), SWIGTYPE_p_Member, SWIG_POINTER_OWN |  0 );
  return resultobj;
fail:
  return NULL;
}
```
The wrapper yields the following compilation error:
```
./vector_mbmvctor_wrap.cxx: In instantiation of ‘SwigValueWrapper<T>& SwigValueWrapper<T>::operator=(const T&) [with T = Member]’:
./vector_mbmvctor_wrap.cxx:4661:10:   required from here
./vector_mbmvctor_wrap.cxx:32:73: error: use of deleted function ‘Member::Member(const Member&)’
 igValueWrapper& operator=(const T& t) { SwigMovePointer tmp(new T(t)); pointer 
```

It could be fixed respecting the constructor semantics via `is_move_constructible()`:
```c++
#if __cplusplus >= 201103L
  #include <type_traits>
  using std::is_move_constructible;
  #define MOVE_SEMANTICS
#endif  // No MOVE_SEMANTICS
...
  result = (arg1)->release();
  resultobj = SWIG_NewPointerObj(new (
#ifdef MOVE_SEMANTICS
    is_move_constructible<Member>::value
    ? Member(move(result)) :
#endif  // MOVE_SEMANTICS
    Member(result)
  ) , SWIGTYPE_p_Member, SWIG_POINTER_OWN |  0 );
...
```

**Comment by luav** [comment]
The manual workaround is to create another member function that
- either returns the value by reference / pointer, retaining the value ownership by the original object,
- or returns created unique/shared_ptr to which the value is moved, transferring the ownership to the caller.

The adapter, which returns the value by reference is preferable in most cases and could be generated automatically when the the value has move-only semantics, for the example see #1043.

**Comment by ojwb** [comment]
Related to #999.

**Comment by luav** [comment]
Note that proposed solution besides solving the issue also would result in much more efficient SWIG wrappers when containers are returned (copy of the whole container is avoided for the temporary stored result, which is required to be able to use typemaps). Moreover, the solution does not affect behavior of the caller at all.

**Comment by ix-dcourtois** [comment]
Hi, I know this issue is pretty old, but is there any plan to integrate the fix suggested by @luav ? I only see advantages to it: more efficient code, still compatible with c++98 since it's protected by macros.. @ojwb ?

**Comment by ojwb** [comment]
It looks like @luav never submitted this as a PR.  I don't think anybody else has picked up the idea so if you want to work on it, go for it.

**Comment by ix-dcourtois** [comment]
@ojwb Thanks for the answer. I'll see if I can do it (I never opened the sources of Swig, I hope the fix will be as "easy" as it appear to implement :p)

**Comment by d-b-w** [comment]
Looks like the first part of the fix is to take advantage of the move constructor in SwigValueWrapper: https://github.com/swig/swig/blob/6f2399e86da13a9feb436e3977e15d2b9738294e/Lib/swig.swg#L675

**Comment by ojwb** [comment]
That's just moving a pointer isn't it?  I'm not seeing how a move constructor on the pointed-to type is going to help there.

**Comment by d-b-w** [comment]
I have a move-only class for which one of the failure messages when I attempt a SWIG wrapping is:

    cpp11_move_only_wrap.cxx: In instantiation of ‘SwigValueWrapper<T>& SwigValueWrapper<T>::operator=(const T&) [with T = MoveOnly]’:
    cpp11_move_only_wrap.cxx:3119:24:   required from here
    cpp11_move_only_wrap.cxx:34:65: error: use of deleted function ‘MoveOnly::MoveOnly(const MoveOnly&)’
    SwigValueWrapper& operator=(const T& t) { SwigMovePointer tmp(new T(t)); pointer = tmp; return *this; }

If I add:

    SwigValueWrapper& operator=(T&& t) { SwigMovePointer tmp(new T(std::move(t))); pointer = tmp; return *this; }

at least one failure mode is addressed. There are still other problems in the wrapping, of course:

    cpp11_move_only_wrap.cxx: In member function ‘MoveOnly& MoveOnly::operator=(MoveOnly&&)’:
    cpp11_move_only_wrap.cxx:2794:14: warning: reference to local variable ‘new_m’ returned [-Wreturn-local-addr]
        MoveOnly new_m(std::move(m));
                ^~~~~
    cpp11_move_only_wrap.cxx: In function ‘PyObject* _wrap_factory(PyObject*, PyObject*)’:
    cpp11_move_only_wrap.cxx:3120:86: error: use of deleted function ‘MoveOnly::MoveOnly(const MoveOnly&)’
    resultobj = SWIG_NewPointerObj((new MoveOnly(static_cast< const MoveOnly& >(result))), SWIGTYPE_p_MoveOnly, SWIG_POINTER_OWN |  0 );

**Comment by ojwb** [comment]
Oh I see - it helps if there is **only** a move constructor.

```
  MoveOnly& operator=(MoveOnly &&m){
    MoveOnly new_m(std::move(m));
    return new_m;
  }
```

A move constructor should always return `*this` shouldn't it?  And this is not assigning to `m_value` of the object it is called on in any way.

**Comment by d-b-w** [comment]
Good point

**Comment by wsfulton** [comment]
I couldn't get the `is_move_constructible` solution to work, I get errors like this:
```
error: use of deleted function ‘constexpr Member::Member(const Member&)
```
as there is still code, that needs compiling, that uses the deleted copy constructor in the ternary expression.

I've put in a simpler solution and removed all the casts we had in the "out" typemaps when copying the type. This allows the compiler to call the appropriate constructor (copy constructor or move constructor) when using C++11 and later. A copy constructor will still be called if there is no move constructor. The move constructor will be called if one is available. It took quite a bit of experimenting to hone in on this solution; in the end the simple code without any casts worked best.

`SwigValueWrapper` was fixed in 088dc6e8707b45872c687c980349ece6866c0bfd.
